etl/steps/data/garden/harvard/2023-09-18/colonial_dates_dataset.py

```diff
@@ -0,0 +1,128 @@
+    tb_rest = tb_rest[~tb_rest["country"].isin(colonized_list + colonizers_list)].reset_index(drop=True)
+
+    # Remove duplicates for tb_rest
+    tb_rest = tb_rest.drop_duplicates(subset=["country"], keep="first").reset_index(drop=True)
+
+    # For these countries, assign the minimum year of colstart_max as colstart_max and the maximum year of colend_max as colend_max
+    tb_rest["colstart_max"] = tb_colonized["colstart_max"].min()
+    tb_rest["colend_max"] = YEAR
+
+    # Create a year column with one value per row representing the range between colstart_max and colend_max
+    # NOTE: I have decided to use last date aggregations, but we could also use mean aggregations
+    tb_colonized["year"] = tb_colonized.apply(lambda x: list(range(x["colstart_max"], x["colend_max"] + 1)), axis=1)
+    tb_rest["year"] = tb_rest.apply(lambda x: list(range(x["colstart_max"], x["colend_max"] + 1)), axis=1)
+
+    # Explode the year column
+    tb_colonized = tb_colonized.explode("year").reset_index(drop=True)
+    tb_rest = tb_rest.explode("year").reset_index(drop=True)
+
+    # Drop colstart and colend columns
+    tb_colonized = tb_colonized.drop(columns=["colstart_max", "colend_max", "colstart_mean", "colend_mean", "col"])
+    tb_rest = tb_rest.drop(columns=["colstart_max", "colend_max", "colstart_mean", "colend_mean", "col"])
+
+    # Create another table with the total number of colonies per colonizer and year
+    tb_count = tb_colonized.groupby(["colonizer", "year"]).agg({"country": "count"}).reset_index().copy_metadata(tb)
+
+    # Rename columns
+    tb_count = tb_count.rename(columns={"colonizer": "country", "country": "total_colonies"})
+
+    # Consolidate results in country and year columns, by merging colonizer column in each row
+    tb_colonized = (
+        tb_colonized.groupby(["country", "year"])
+        .agg({"colonizer": lambda x: " - ".join(x)})
+        .reset_index()
+        .copy_metadata(tb)
+    )
+
+    # Concatenate tb_colonized and tb_rest
+    tb = pr.concat([tb_colonized, tb_rest, tb_count], short_name="colonial_dates_dataset")
+
+    # Fill years in the range (tb_colonized['year'].min(), YEAR) not present for each country
+    tb = tb.set_index(["country", "year"]).unstack().stack(dropna=False).reset_index()
+
+    # Create an additional summarized colonizer column, replacing the values with " - " with "More than one colonizer"
+    # I add the "z." to have this at the last position of the map brackets
+    tb["colonizer_grouped"] = tb["colonizer"].apply(
+        lambda x: "z. Multiple colonizers" if isinstance(x, str) and " - " in x else x
+    )
+
+    # Create last_colonizer column, which is the most recent non-null colonizer for each country and year
+    tb["last_colonizer"] = tb.groupby(["country"])["colonizer"].fillna(method="ffill")
+    tb["last_colonizer_grouped"] = tb.groupby(["country"])["colonizer_grouped"].fillna(method="ffill")
```

Comment:
I see metadata is not kept when using this and not what it is commented, @Marigold. Am I doing something wrong?

Reply:
Thanks for spotting this! It turned out that `fillna` behaves differently from other groupby functions. I pushed fix to master and rebased this branch (you can get the changes with `git fetch --all && git reset --hard origin/fix-create-colonial-dates-dataset`)